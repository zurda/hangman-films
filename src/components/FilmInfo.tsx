import React from 'react'
import styled from '@emotion/styled'

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  font-family: 'Roboto';
  @media (min-width: 420px) {
    flex-direction: row;
    justify-content: space-between;
  }
`

const Info = styled.p`
  margin: 16px 32px;
`

const Img = styled.img`
  width: 154px;
`

const FilmInfo = ({ film }: any) => (
  <Container>
    <Img
      src={`http://image.tmdb.org/t/p/w154${film.poster_path}`}
      alt="Film poster"
    />
    {film.overview || film.tagline ? (
      <Info>{film.overview || film.tagline}</Info>
    ) : null}
  </Container>
)

export default FilmInfo
