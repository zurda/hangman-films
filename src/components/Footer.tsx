import React from 'react'
import styled from '@emotion/styled'
import { colours } from '../style/shared'

const Container = styled.footer`
  width: 100%;
  color: ${colours.diamond};
  background-color: ${colours.pictonBlue};
  font-weight: bold;
  font-family: 'Courier New', Courier, monospace;
  padding: 16px 32px;
  text-align: center;
`

const Description = styled.p`
  margin: 0;
  & a {
    color: inherit;
  }
`

const Footer = () => (
  <Container>
    <Description>
      This project is available on{' '}
      <a href="https://github.com/zurda/hangman-react">Github</a>
    </Description>
  </Container>
)

export default Footer
