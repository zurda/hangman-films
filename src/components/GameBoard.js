import React from 'react'
import styled from '@emotion/styled'
import { colours } from '../style/shared'

const Container = styled.div`
  width: 100%;
  color: ${colours.maastrichtBlue};
  padding: 16px 32px;
  background-color: ${colours.lightest};
  min-height: calc(100vh - 130px);
  @media (min-width: 420px) {
    font-size: 18px;
  }
`

const ContentContainer = styled.div`
  max-width: 800px;
  margin: 0 auto;
`

const GameBoard = ({ children }) => (
  <Container>
    <ContentContainer>{children}</ContentContainer>
  </Container>
)

export default GameBoard
