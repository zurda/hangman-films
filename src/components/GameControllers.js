import React from 'react'
import styled from '@emotion/styled'
import { colours } from '../style/shared'
import downArrowSvg from '../style/assets/down_arrow.svg'

export const Button = styled.button`
  color: ${(props) => (props.type === 'primary' ? 'white' : 'white')};
  background-color: ${(props) =>
    props.type === 'primary' ? colours.eucaliptus : colours.iceberg};
  font-weight: bold;
  padding: 8px 16px;
  text-align: center;
  -webkit-transition-duration: 0.2s; /* Safari */
  transition-duration: 0.2s;
  border: none;
  width: 100%;
  margin: 8px 0;
  box-shadow: 0px 3px 15px rgba(0, 0, 0, 0.2);
  border-radius: 2px;
  &:hover {
    background-color: ${colours.maastrichtBlue};
    color: ${colours.lightest};
  }
  @media (min-width: 420px) {
    font-size: 18px;
  }
`

const Text = styled.span`
  width: 200px;
`

const DropdownContainer = styled.div`
  display: flex;
  flex-direction: row;
  padding: 0;
  margin: 0;
`

const Select = styled.select`
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: white;
  background-image: url(${downArrowSvg});
  background-size: 8px;
  background-repeat: no-repeat;
  background-position-y: 50%;
  background-position-x: calc(100% - 8px);
  text-align: left;
  padding: 8px;
  margin-bottom: 8px;
  cursor: pointer;
  width: 100%;
  border: 1px solid black;
  border: none;
  border-radius: 2px;
  box-shadow: 0px 3px 15px rgba(0, 0, 0, 0.2);
  @media (min-width: 420px) {
    font-size: 18px;
  }
`

export const Dropdown = ({ onClick, onChange, title, options }) => (
  <DropdownContainer>
    {title && title.length > 0 && <Text>{title}</Text>}
    <Select onClick={() => onClick('')} onChange={(e) => onChange(e.target.value)}>
      <option value={''}>Choose a genre:</option>
      {options.map((option, index) => (
        <option key={index} value={option.name}>
          {option.name}
        </option>
      ))}
    </Select>
  </DropdownContainer>
)
