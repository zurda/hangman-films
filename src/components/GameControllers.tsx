import React from 'react'
import styled from '@emotion/styled'
import { colours } from '../style/shared'
import downArrowSvg from '../style/assets/down_arrow.svg'

export const Button = styled.button`
  color: white;
  background-color: ${colours.iceberg};
  font-weight: bold;
  padding: 8px 16px;
  text-align: center;
  -webkit-transition-duration: 0.2s; /* Safari */
  transition-duration: 0.2s;
  border: none;
  width: 100%;
  box-shadow: 0px 3px 15px rgba(0, 0, 0, 0.2);
  border-radius: 2px;
  &:hover {
    background-color: ${colours.maastrichtBlue};
    color: ${colours.lightest};
  }
  @media (min-width: 420px) {
    font-size: 18px;
  }
`

export const RoundButton = styled(Button)`
  border-radius: 50%;
  width: 80px;
  height: 80px;
  align-self: flex-end;
  margin: 0 auto;
`

export const ButtonContainer = styled.div`
  width: 100%;
  @media (min-width: 420px) {
    width: 49%;
  }
`

export const FlexResponsive = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  @media (min-width: 420px) {
    flex-direction: row;
    justify-content: space-between;
  }
`

const Text = styled.span`
  width: 200px;
`

const DropdownContainer = styled.div`
  display: flex;
  flex-direction: row;
  padding: 0;
  margin: 0;
  width: 100%;
  padding-bottom: 8px;
  @media (min-width: 420px) {
    width: 49%;
    padding-bottom: 0;
  }
`

const Select = styled.select`
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: white;
  background-image: url(${downArrowSvg});
  background-size: 8px;
  background-repeat: no-repeat;
  background-position-y: 50%;
  background-position-x: calc(100% - 8px);
  text-align: left;
  padding: 8px;
  cursor: pointer;
  width: 100%;
  border: 1px solid black;
  border: none;
  border-radius: 2px;
  box-shadow: 0px 3px 15px rgba(0, 0, 0, 0.2);
  @media (min-width: 420px) {
    font-size: 18px;
  }
`

interface IOption {
  name: string
}

interface IDropdown {
  options: IOption[]
  onClick: any
  onChange: any
  title: string
}

export const Dropdown = ({ onClick, onChange, title, options }: IDropdown) => (
  <DropdownContainer>
    {title && title.length > 0 && <Text>{title}</Text>}
    <Select onClick={() => onClick('')} onChange={(e) => onChange(e.target.value)}>
      <option value={''}>Choose a genre:</option>
      {options.map((option, index) => (
        <option key={index} value={option.name}>
          {option.name}
        </option>
      ))}
    </Select>
  </DropdownContainer>
)
