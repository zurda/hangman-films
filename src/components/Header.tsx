import React from 'react'
import styled from '@emotion/styled'
import { colours } from '../style/shared'

const Container = styled.header`
  width: 100%;
  background-color: ${colours.maastrichtBlue};
  color: ${colours.diamond};
  font-weight: bold;
  font-family: 'Courier New', Courier, monospace;
  padding: 16px 32px;
`

const Heading = styled.h1`
  margin: 0;
`

const Description = styled.p`
  margin: 0;
  color: ${colours.pictonBlue};
`

interface IHeaderProps {
  appName: string
  appDescription: string
}

const Header = ({ appName, appDescription }: IHeaderProps) => (
  <Container>
    <Heading>{appName}</Heading>
    <Description>{appDescription}</Description>
  </Container>
)

export default Header
