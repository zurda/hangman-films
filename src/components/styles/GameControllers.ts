import styled from 'styled-components';
import downArrowSvg from '../../style/assets/down_arrow.svg';

export const DropdownContainer = styled.div`
  flex:1;

  background-color: ${(props) => props.theme.colors.light};

  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 0;
  margin: 0;

  height: 40px;
  border-radius: 8px;

  select {
    color: ${(props) => props.theme.colors.text};
    

    &:focus {
      border: 2px solid ${(props) => props.theme.colors.text};
    }
  }

  @media (max-width: 420px) {
    width: 100%;
  }
`

export const Select = styled.select`
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: ${(props) => props.theme.colors.light};;
  background-image: url(${downArrowSvg});
  background-size: 8px;
  background-repeat: no-repeat;
  background-position-y: 50%;
  background-position-x: calc(100% - 8px);
  border-radius: 8px;

  width: 100%;
  text-align: left;
  padding: 8px;
  cursor: pointer;
  border: none;

  font-size: 18px;
  font-weight: 500;
`