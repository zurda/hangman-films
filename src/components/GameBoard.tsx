import React, { useState, useEffect, useMemo, useCallback } from 'react'

import { allLetters, getGenreId, defaultLetters, getRandom } from '../helpers'
import { fetchGenres, fetchFilm } from '../api'

import { useGameData } from '../hooks/GameData'

import HiddenText from '../components/HiddenText'
import LettersTray from '../components/LettersTray'
import { Dropdown } from '../components/GameControllers'
import ReleaseHint from '../components/ReleaseHint'

import { FaVideo } from 'react-icons/fa'

import {
  Container,
  ContentContainer,
  GenreCard,
  Button,
  ButtonContainer,
  HiddenTextContainer,
  LettersTrayContainer,
  FilmContainer,
  ImageContainer,
  HintButton,
  FilmInfo,
  PosterOverlay,
  Text,
  PosterContainer
} from './styles/GameBoard'

interface FilmInfo {
  id: string
  title: string
  poster_path: string
  overview: string
  tagline: string
  release_date: string
}

const EMPTY_FILM: FilmInfo = {
  id: '',
  title: '',
  poster_path: '',
  overview: '',
  tagline: '',
  release_date: ''
}

const MAX_HINT_COUNT = 2
const INITIAL_HINT_COST = 2
const MIN_HINT_COST = 1

const GameBoard = () => {
  const { alreadyPlayed, saveMovieId, numberOfLives } = useGameData()

  const [counter, setCounter] = useState(numberOfLives)
  const [film, setFilm] = useState(EMPTY_FILM)
  const [genres, setGenres] = useState<any[]>([])
  const [selectedGenre, setSelectedGenre] = useState({ id: 99 })
  const [guessedLetters, setGuessedLetters] = useState<string[]>(defaultLetters)
  const [hintCounter, setHintCounter] = useState(0)
  const [gameWon, setGameWon] = useState<boolean>(false)
  const [posterOverlay, setPosterOverlay] = useState(
    Array.from({ length: 28 }, () => true)
  )

  const updateCounter = (amount: number) => {
    if (counter + amount < 1) {
      const newPosterOverlay = posterOverlay.map((_) => false)
      setPosterOverlay(newPosterOverlay)
    }
    setCounter(counter + amount)
  }

  const onCharClick = (char: string) => {
    setGuessedLetters([...guessedLetters, char, char.toLowerCase()])
    return film.title.indexOf(char) > -1 ||
      film.title.indexOf(char.toLowerCase()) > -1
      ? null
      : updateCounter(-1)
  }

  const resetState = useCallback(() => {
    setFilm(EMPTY_FILM)
    setGuessedLetters(defaultLetters)
    setCounter(numberOfLives)
    setHintCounter(0)
    setGameWon(false)

    const flushedOverlay = posterOverlay.map(() => true)
    setPosterOverlay(flushedOverlay)
  }, [numberOfLives, posterOverlay])

  const onGetFilmClick = async () => {
    resetState()
    const genreId = getGenreId(selectedGenre, genres)
    const filmId = await fetchFilm(genreId, alreadyPlayed, setFilm)
    saveMovieId(filmId)
  }

  const onHintClick = () => {
    updateCounter(
      -Math.max(MIN_HINT_COST, Math.abs(hintCounter - INITIAL_HINT_COST))
    )

    const newPosterOverlay = posterOverlay
    const availableIndexes = newPosterOverlay.flatMap((item, index) =>
      item ? [index] : []
    )
    const randomAvailableIndex = availableIndexes[getRandom(availableIndexes.length)]

    newPosterOverlay[randomAvailableIndex] = false

    setHintCounter(hintCounter + 1)

    setPosterOverlay(newPosterOverlay)
  }
  const hearts = useMemo(() => {
    return Array.from({ length: counter }, (_, index) => ({ id: `Heart-${index}` }))
  }, [counter])

  useEffect(() => {
    fetchGenres().then((result) => setGenres(result.genres))
  }, [])
  useEffect(() => {
    if (counter < 1) {
      setGuessedLetters(allLetters)
    }
    const remainingLetters = film.title
      .split('')
      .filter((filmLetter) => guessedLetters.indexOf(filmLetter) === -1)
    if (film.title !== '' && remainingLetters.length === 0) {
      setGuessedLetters(allLetters)
      setGameWon(true)
    }
  }, [film.title, guessedLetters, counter])

  useEffect(() => {
    setFilm(EMPTY_FILM)
    setGuessedLetters(defaultLetters)
    setCounter(numberOfLives)
    setHintCounter(0)

    const flushedOverlay = Array.from({ length: posterOverlay.length }, () => true)
    setPosterOverlay(flushedOverlay)
  }, [numberOfLives, posterOverlay.length])

  return (
    <Container>
      <ContentContainer>
        <GenreCard>
          {genres && genres.length > 0 && (
            <Dropdown
              options={genres}
              title=""
              onClick={() => setFilm(EMPTY_FILM)}
              onChange={setSelectedGenre}
            />
          )}
          <ButtonContainer>
            <Button onClick={onGetFilmClick}>Get film</Button>
          </ButtonContainer>
        </GenreCard>
        {film.title !== '' && (
          <>
            <HiddenTextContainer>
              <HiddenText filmTitle={film.title} guessedLetters={guessedLetters} />
            </HiddenTextContainer>

            <LettersTrayContainer>
              <Text>
                {counter > 0 && (
                  <>
                    {gameWon ? (
                      `You won!üéâ\nWith ${hearts.length} guess${
                        hearts.length > 1 ? 'es' : ''
                      } to go üëè\n`
                    ) : (
                      <>
                        Guesses left:
                        {hearts.map((item) => (
                          <FaVideo key={item.id} />
                        ))}
                      </>
                    )}
                  </>
                )}
              </Text>
              <LettersTray
                guessedLetters={guessedLetters}
                onClickHandler={onCharClick}
              />
            </LettersTrayContainer>

            <FilmContainer style={{ justifyContent: 'flex-start' }}>
              <PosterContainer>
                <ImageContainer posterImage={film.poster_path}>
                  {!gameWon &&
                    posterOverlay.map((status, index) => (
                      <PosterOverlay key={`Overlay-${index}`} active={!status} />
                    ))}
                </ImageContainer>
                {counter > MAX_HINT_COUNT &&
                  hintCounter < MAX_HINT_COUNT &&
                  !gameWon && (
                    <HintButton type="button" onClick={onHintClick}>
                      <h4>HINT</h4>
                      <p>
                        This will cost you{' '}
                        {Math.abs(hintCounter - INITIAL_HINT_COST)} guess(es)!
                      </p>
                    </HintButton>
                  )}
              </PosterContainer>

              {(counter < 1 || gameWon) && (
                <FilmInfo>
                  <h3>{film.title}</h3>
                  {film.overview || film.tagline ? (
                    <p>{film.overview || film.tagline}</p>
                  ) : null}
                </FilmInfo>
              )}
              <ReleaseHint
                releaseYear={film.release_date.substring(0, 4)}
                counter={counter}
                setCounter={setCounter}
                gameWon={gameWon}
              />
            </FilmContainer>
          </>
        )}
      </ContentContainer>
    </Container>
  )
}

export default GameBoard
