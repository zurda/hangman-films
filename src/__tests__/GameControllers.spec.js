import React from 'react'
import { create } from 'react-test-renderer'
import { Button, RoundButton, Dropdown } from '../components/GameControllers'

const options = [{ name: 'First option' }, { name: 'second option' }]
const onClick = () => {}
const onChange = () => {}

describe('GameControllers component', () => {
  test('RoundButton matches the snapshot', () => {
    const roundButton = create(<RoundButton>Button name</RoundButton>)
    expect(roundButton.toJSON()).toMatchSnapshot()
  })
  test('Button matches the snapshot', () => {
    const button = create(<Button>Button name</Button>)
    expect(button.toJSON()).toMatchSnapshot()
  })
  test('Dropdown matches the snapshot', () => {
    const dropdown = create(
      <Dropdown options={options} onChange={onChange} onClick={onClick} />
    )
    expect(dropdown.toJSON()).toMatchSnapshot()
  })
})
