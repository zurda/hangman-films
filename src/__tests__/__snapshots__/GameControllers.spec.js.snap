// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameControllers component Button matches the snapshot 1`] = `
<button
  className="css-14nymm2"
>
  Button name
</button>
`;

exports[`GameControllers component Dropdown matches the snapshot 1`] = `
<div
  className="css-1t4fj7x"
>
  <select
    className="css-zjhfov"
    onChange={[Function]}
    onClick={[Function]}
  >
    <option
      value=""
    >
      Choose a genre:
    </option>
    <option
      value="First option"
    >
      First option
    </option>
    <option
      value="second option"
    >
      second option
    </option>
  </select>
</div>
`;

exports[`GameControllers component RoundButton matches the snapshot 1`] = `
<button
  className="css-15j8xbs"
>
  Button name
</button>
`;
